File: streamlit_template/sample_data.py
"""Generate sample data for demonstration purposes"""

import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import random

def generate_sample_ohlcv(symbol, days=365):
    """Generate sample OHLCV data for a symbol"""
    np.random.seed(hash(symbol) % 2**32)  # Consistent data per symbol
    
    # Base prices for different symbols
    base_prices = {
        'BTCUSDT': 45000, 'ETHUSDT': 3000, 'BNBUSDT': 400, 'ADAUSDT': 0.5,
        'XRPUSDT': 0.6, 'SOLUSDT': 100, 'DOTUSDT': 25, 'DOGEUSDT': 0.08,
        'AVAXUSDT': 35, 'SHIBUSDT': 0.00002, 'MATICUSDT': 1.2, 'LTCUSDT': 150,
        'UNIUSDT': 25, 'LINKUSDT': 15, 'ATOMUSDT': 12
    }
    
    base_price = base_prices.get(symbol, 100)
    dates = pd.date_range(start=datetime.now() - timedelta(days=days), 
                         end=datetime.now(), freq='D')
    
    # Generate price movements with trend
    price_changes = np.random.normal(0, 0.02, len(dates))  # 2% daily volatility
    trend = np.linspace(-0.1, 0.3, len(dates))  # Overall upward trend
    price_changes += trend * 0.001  # Add trend component
    
    prices = [base_price]
    for change in price_changes[1:]:
        new_price = prices[-1] * (1 + change)
        prices.append(max(new_price, base_price * 0.1))  # Prevent extreme drops
    
    # Generate OHLCV data
    data = []
    for i, (date, close) in enumerate(zip(dates, prices)):
        daily_vol = abs(np.random.normal(0, 0.015))  # Daily volatility
        high = close * (1 + daily_vol)
        low = close * (1 - daily_vol)
        open_price = prices[i-1] if i > 0 else close
        volume = random.randint(1000000, 50000000)
        
        data.append({
            'Date': date,
            'Open': open_price,
            'High': high,
            'Low': low,
            'Close': close,
            'Volume': volume
        })
    
    return pd.DataFrame(data)

def generate_sample_trades(symbol, num_trades=50):
    """Generate sample trading history"""
    trades = []
    base_price = 100
    
    for i in range(num_trades):
        entry_date = datetime.now() - timedelta(days=random.randint(1, 365))
        exit_date = entry_date + timedelta(days=random.randint(1, 30))
        
        entry_price = base_price * (1 + random.uniform(-0.2, 0.2))
        exit_price = entry_price * (1 + random.uniform(-0.1, 0.15))
        
        quantity = random.uniform(0.1, 10)
        side = random.choice(['BUY', 'SELL'])
        
        pnl = (exit_price - entry_price) * quantity if side == 'BUY' else (entry_price - exit_price) * quantity
        
        trades.append({
            'Symbol': symbol,
            'Side': side,
            'Entry_Date': entry_date,
            'Exit_Date': exit_date,
            'Entry_Price': round(entry_price, 4),
            'Exit_Price': round(exit_price, 4),
            'Quantity': round(quantity, 4),
            'PnL': round(pnl, 2),
            'Status': 'Closed'
        })
    
    return pd.DataFrame(trades)

def generate_pattern_signals():
    """Generate sample pattern detection signals"""
    patterns = [
        'Rising Wedge', 'Falling Wedge', 'Ascending Triangle', 'Descending Triangle',
        'Head and Shoulders', 'Inverse Head and Shoulders', 'Bull Flag', 'Bear Flag',
        'Double Top', 'Double Bottom', 'Cup and Handle', 'Symmetrical Triangle'
    ]
    
    signals = []
    for i in range(20):
        signal = {
            'Symbol': random.choice(['BTCUSDT', 'ETHUSDT', 'BNBUSDT', 'ADAUSDT']),
            'Pattern': random.choice(patterns),
            'Signal': random.choice(['Buy', 'Sell', 'Strong Buy', 'Strong Sell', 'Neutral']),
            'Confidence': random.randint(60, 95),
            'Timeframe': random.choice(['1h', '4h', '1d', '1w']),
            'Detected_At': datetime.now() - timedelta(hours=random.randint(1, 72))
        }
        signals.append(signal)
    
    return pd.DataFrame(signals)

def generate_market_sentiment():
    """Generate sample market sentiment data"""
    return {
        'Fear_Greed_Index': random.randint(20, 80),
        'Social_Sentiment': random.uniform(-1, 1),
        'News_Sentiment': random.uniform(-1, 1),
        'Technical_Sentiment': random.uniform(-1, 1),
        'Overall_Sentiment': random.choice(['Bullish', 'Bearish', 'Neutral'])
    }

def generate_cycle_analysis(symbol):
    """Generate sample cycle analysis"""
    phases = ['Early Bull', 'Mid Bull', 'Late Bull', 'Early Bear', 'Mid Bear', 'Late Bear']
    current_phase = random.choice(phases)
    duration = random.randint(10, 90)
    
    return {
        'symbol': symbol,
        'current_phase': current_phase,
        'duration': duration,
        'momentum': random.uniform(0, 1),
        'bull_probability': random.randint(30, 90),
        'bear_probability': random.randint(10, 70),
        'risk_level': random.choice(['Low', 'Medium', 'High']),
        'expected_continuation': f"{random.randint(5, 60)} days"
    }

File: src/utils/sampleData.js
// Sample data generation for the trading dashboard

export const generateSampleData = (symbol = 'BTCUSDT') => {
  const data = []
  const basePrice = getBasePrice(symbol)
  let currentPrice = basePrice
  
  for (let i = 0; i < 100; i++) {
    const date = new Date()
    date.setDate(date.getDate() - (99 - i))
    
    // Generate realistic price movement
    const volatility = 0.02 // 2% daily volatility
    const trend = Math.sin(i / 20) * 0.001 // Slight trend
    const randomChange = (Math.random() - 0.5) * volatility + trend
    
    currentPrice = currentPrice * (1 + randomChange)
    
    // Generate OHLC data
    const high = currentPrice * (1 + Math.random() * 0.01)
    const low = currentPrice * (1 - Math.random() * 0.01)
    const open = i === 0 ? basePrice : data[i-1]?.close || currentPrice
    
    data.push({
      date: date.toISOString().split('T')[0],
      open: parseFloat(open.toFixed(4)),
      high: parseFloat(high.toFixed(4)),
      low: parseFloat(low.toFixed(4)),
      close: parseFloat(currentPrice.toFixed(4)),
      volume: Math.floor(Math.random() * 1000000) + 100000
    })
  }
  
  return data
}

export const generateMarketData = () => {
  const symbols = ['BTCUSDT', 'ETHUSDT', 'BNBUSDT', 'ADAUSDT', 'XRPUSDT', 'SOLUSDT', 'DOGEUSDT', 'MATICUSDT']
  
  return symbols.map(symbol => {
    const basePrice = getBasePrice(symbol)
    const change = (Math.random() - 0.5) * 20 // -10% to +10%
    const price = basePrice * (1 + change / 100)
    
    return {
      symbol,
      price: parseFloat(price.toFixed(4)),
      change: parseFloat(change.toFixed(2)),
      volume: Math.floor(Math.random() * 1000000000) + 10000000
    }
  })
}

export const generateSignals = () => {
  const symbols = ['BTCUSDT', 'ETHUSDT', 'BNBUSDT', 'ADAUSDT']
  const signals = ['Strong Buy', 'Buy', 'Neutral', 'Sell', 'Strong Sell']
  const patterns = [
    'Rising Wedge', 'Falling Wedge', 'Head & Shoulders', 'Bull Flag', 
    'Bear Flag', 'Double Top', 'Double Bottom', 'Ascending Triangle'
  ]
  const timeframes = ['1h', '4h', '1d', '1w']
  
  return symbols.slice(0, 3).map(symbol => ({
    symbol,
    signal: signals[Math.floor(Math.random() * signals.length)],
    confidence: Math.floor(Math.random() * 40) + 60, // 60-100%
    pattern: patterns[Math.floor(Math.random() * patterns.length)],
    timeframe: timeframes[Math.floor(Math.random() * timeframes.length)],
    timestamp: new Date(Date.now() - Math.random() * 3600000).toISOString() // Last hour
  }))
}

const getBasePrice = (symbol) => {
  const basePrices = {
    'BTCUSDT': 45000,
    'ETHUSDT': 3000,
    'BNBUSDT': 400,
    'ADAUSDT': 0.5,
    'XRPUSDT': 0.6,
    'SOLUSDT': 100,
    'DOGEUSDT': 0.08,
    'MATICUSDT': 1.2
  }
  
  return basePrices[symbol] || 100
}

export const generatePatternData = () => {
  return {
    patterns: [
      { name: 'Rising Wedge', detected: true, confidence: 85, type: 'bearish' },
      { name: 'Bull Flag', detected: true, confidence: 72, type: 'bullish' },
      { name: 'Ascending Triangle', detected: true, confidence: 91, type: 'bullish' },
    ],
    smartMoney: [
      { name: 'Order Block', status: 'Bullish', level: '$45,200' },
      { name: 'Fair Value Gap', status: 'Filled', level: '$44,800' },
      { name: 'Liquidity Zone', status: 'Active', level: '$46,500' },
    ]
  }
}